### PR TITLE
NextJS: Allow disabling next/image lazy loading

### PR DESCRIPTION
**Pull request by martinnabhan**
## What I did

`next/image` lazy loads images by default. When using Storybook (and especially Chromatic) we usually want images below the fold to load as well so they show up in snapshots. This pull request sets `next/image`'s `loading` property to `eager` which disables lazy loading.

## How to test

1. Create a story using `next/image` and place an image below the fold
2. Open Storybook in your browser
3. Access the story and make sure the image gets loaded even if it's below the fold (or use Chromatic to snapshot the story and make sure images below the fold are loaded and visible)

## Checklist

- [x] Make sure your changes are tested (stories and/or unit, integration, or end-to-end tests)
- [x] Make sure to add/update documentation regarding your changes
- [x] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

#### Maintainers

- [ ] If this PR should be tested against many or all sandboxes,
      make sure to add the `ci:merged` or `ci:daily` GH label to it.
- [ ] Make sure this PR contains **one** of the labels below.

`["cleanup", "BREAKING CHANGE", "feature request", "bug", "documentation", "maintenance", "dependencies", "other"]`

### CODE REVIEW

**Comment by valentinpalkovic** [comment]
@martinnabhan Thank you so much for your contribution.

Your PR would introduce a breaking change where people might rely on lazy loading behavior. Instead, this option should be configurable. Let's add it to the `parameters.nextjs` object instead, where people like you could define the default loading behaviour.

**Comment by martinnabhan** [comment]
I thought that might be the way to go! I'll see if I can turn it into a parameter some time this week.

File: code/e2e-tests/framework-nextjs.spec.ts [hunk 1]
```diff
@@ -21,6 +21,30 @@ test.describe('Next.js', () => {
...
+    test.beforeEach(async ({ page }) => {
+      sbPage = new SbPage(page);
+    });
+
+    test('should lazy load images by default', async () => {
+      await sbPage.navigateToStory('frameworks/nextjs/Image', 'lazy');
+
+      const img = sbPage.previewRoot().locator('img');
+
+      expect(await img.evaluate<boolean, HTMLImageElement>((image) => image.complete)).toBeFalsy();
```
Comment (martinnabhan) [inline code/e2e-tests/framework-nextjs.spec.ts:36]: A lazy loaded image element's complete property will be false if the image hasn't been loaded yet.

File: code/frameworks/nextjs/src/images/decorator.tsx [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+): React.ReactNode => {
+  if (!parameters.nextjs?.image) {
+    return <Story />;
+  }
+
+  return (
+    <ImageContext.Provider value={parameters.nextjs.image}>
+      <Story />
+    </ImageContext.Provider>
+  );
```
Comment (martinnabhan) [inline code/frameworks/nextjs/src/images/decorator.tsx:17]: I tried other methods such as using environment variables or traversing the React element tree and adding the loading property, but in the end using Context seemed to be the most straightforward approach.

File: code/frameworks/nextjs/src/images/next-image-stub.tsx [hunk 3]
```diff
@@ -38,7 +39,11 @@ const OriginalNextImage = NextImage.default;
 Object.defineProperty(NextImage, 'default', {
   configurable: true,
   value: (props: _NextImage.ImageProps) => {
-    return <OriginalNextImage {...props} loader={props.loader ?? defaultLoader} />;
+    const imageParameters = React.useContext(ImageContext);
+
+    return (
+      <OriginalNextImage {...props} loader={props.loader ?? defaultLoader} {...imageParameters} />
```
Comment (martinnabhan) [inline code/frameworks/nextjs/src/images/next-image-stub.tsx:45]: We want passed parameters to override any other props, which is why we spread `imageParameters` last.

File: code/frameworks/nextjs/template/stories/Image.stories.jsx [hunk 4]
```diff
@@ -48,3 +48,30 @@ export const Sized = {
...
+export const Lazy = {
+  args: {
+    src: 'https://storybook.js.org/images/placeholders/50x50.png',
+    width: 50,
+    height: 50,
+  },
+  decorators: [
+    (Story) => (
+      <>
+        <div style={{ height: '200vh' }} />
```
Comment (martinnabhan) [inline code/frameworks/nextjs/template/stories/Image.stories.jsx:61]: To make sure lazy image's don't get loaded I added a `200vh` tall div. The point where a lazy loaded image gets loaded is different for every browser, but `200vh` works for both Chrome and Safari.

File: code/frameworks/nextjs/template/stories/Image.stories.jsx [hunk 5]
```diff
@@ -48,3 +48,30 @@ export const Sized = {
...
+  ],
+};
+
+export const Eager = {
+  ...Lazy,
+  parameters: {
+    nextjs: {
+      image: {
+        loading: 'eager',
+      },
```
Comment (martinnabhan) [inline code/frameworks/nextjs/template/stories/Image.stories.jsx:74]: By passing `loading: 'eager'` we tell `next/image` to eager load every instance of `next/image`.

File: code/frameworks/nextjs/src/images/context.ts [hunk 6]
```diff
@@ -0,0 +1,16 @@
+import { createContext } from 'react';
+import type { ImageProps, StaticImageData } from 'next/image';
+import type { ImageProps as LegacyImageProps } from 'next/legacy/image';
+
+// StaticRequire needs to be in scope for the TypeScript compiler to work.
+// See: https://github.com/microsoft/TypeScript/issues/5711
+// Since next/image doesn't export StaticRequire we need to re-define it here and set src's type to it.
```
Comment (martinnabhan) [inline code/frameworks/nextjs/src/images/context.ts:7]: https://app.circleci.com/pipelines/github/storybookjs/storybook/46783/workflows/1af13e69-d40a-4c2a-8a31-1bb3ad3c8401/jobs/511829

I've never seen this error before but putting StaticRequire in the scope fixed it.

File: code/frameworks/nextjs/README.md [hunk 7]
```diff
@@ -159,12 +159,16 @@ export default {
...
+      image: {
+        loading: 'eager',
+      },
       nextConfigPath: path.resolve(__dirname, '../next.config.js'),
     },
   },
 };
 ```
 
+- `image`: Props to pass to every instance of `next/image`
```
Comment (martinnabhan) [inline code/frameworks/nextjs/README.md:171]: This pull request was originally supposed to just allow toggling lazy loading for every instance of `next/image` using parameters, but I thought we might as well allow setting any property.

File: code/frameworks/nextjs/src/images/next-image-stub.tsx (hunk 3 above)
Comment (valentinpalkovic) [reply code/frameworks/nextjs/src/images/next-image-stub.tsx:45]: As a user I would rather assume, that `imageParameters` are the defaults and I am still able to override the default case by case if I want.

```
<OriginalNextImage {...imageParameters} {...props} loader={props.loader ?? defaultLoader} />
```

File: code/frameworks/nextjs/src/images/next-image-stub.tsx [hunk 8]
```diff
@@ -48,9 +53,17 @@ if (semver.satisfies(process.env.__NEXT_VERSION!, '^13.0.0')) {
...
-      <OriginalNextLegacyImage {...props} loader={props.loader ?? defaultLoader} />
-    ),
+    value: (props: _NextLegacyImage.ImageProps) => {
+      const imageParameters = React.useContext(ImageContext);
+
+      return (
+        <OriginalNextLegacyImage
+          {...props}
+          loader={props.loader ?? defaultLoader}
+          {...imageParameters}
```
Comment (valentinpalkovic) [inline code/frameworks/nextjs/src/images/next-image-stub.tsx:63]: same here

File: code/frameworks/nextjs/src/images/next-image-stub.tsx [hunk 9]
```diff
@@ -60,8 +73,16 @@ if (semver.satisfies(process.env.__NEXT_VERSION!, '^12.2.0')) {
...
-      <OriginalNextFutureImage {...props} loader={props.loader ?? defaultLoader} />
-    ),
+    value: (props: _NextImage.ImageProps) => {
+      const imageParameters = React.useContext(ImageContext);
+
+      return (
+        <OriginalNextFutureImage
+          {...props}
+          loader={props.loader ?? defaultLoader}
+          {...imageParameters}
```
Comment (valentinpalkovic) [inline code/frameworks/nextjs/src/images/next-image-stub.tsx:83]: same here

File: code/frameworks/nextjs/src/images/next-image-stub.tsx (hunk 3 above)
Comment (martinnabhan) [reply code/frameworks/nextjs/src/images/next-image-stub.tsx:45]: I think you're right. Treating `imageParameters` as the defaults makes sense to me!

**Review by valentinpalkovic — APPROVED** [review]

**Comment by martinnabhan** [comment]
@valentinpalkovic Thank you for approving! What's the status on this? `build-sandboxes` seems to be failing, but it doesn't seem to be related to this pull request.

**Pull request merged**